src/github/pullRequestModel.ts

```diff
@@ -1751,42 +1752,44 @@ export class PullRequestModel extends IssueModel<PullRequest> implements IPullRe
 		}
 	}
 
-	async markFileAsViewed(filePathOrSubpath: string, event: boolean): Promise<void> {
-		const { mutate, schema } = await this.githubRepository.ensure();
-		const fileName = filePathOrSubpath.startsWith(this.githubRepository.rootUri.path) ?
-			filePathOrSubpath.substring(this.githubRepository.rootUri.path.length + 1) : filePathOrSubpath;
-		await mutate<void>({
-			mutation: schema.MarkFileAsViewed,
-			variables: {
-				input: {
-					path: fileName,
-					pullRequestId: this.graphNodeId,
-				},
-			},
-		});
+	private readonly BATCH_SIZE = 100;
 
-		this.setFileViewedState(fileName, ViewedState.VIEWED, event);
-	}
+	async markFiles(filePathOrSubpaths: string[], event: boolean, state: 'viewed' | 'unviewed'): Promise<void> {
+		const { mutate } = await this.githubRepository.ensure();
+		const pullRequestId = this.graphNodeId;
 
-	async unmarkFileAsViewed(filePathOrSubpath: string, event: boolean): Promise<void> {
-		const { mutate, schema } = await this.githubRepository.ensure();
-		const fileName = filePathOrSubpath.startsWith(this.githubRepository.rootUri.path) ?
-			filePathOrSubpath.substring(this.githubRepository.rootUri.path.length + 1) : filePathOrSubpath;
-		await mutate<void>({
-			mutation: schema.UnmarkFileAsViewed,
-			variables: {
-				input: {
-					path: fileName,
-					pullRequestId: this.graphNodeId,
-				},
-			},
-		});
+		const allFilenames = filePathOrSubpaths
+			.map((f) =>
+				f.startsWith(this.githubRepository.rootUri.path)
+					? f.substring(this.githubRepository.rootUri.path.length + 1)
+					: f
+			);
+
+		const mutationName = state === 'viewed'
+			? 'unmarkFileAsViewed'
+			: 'markFileAsViewed';
+
+
+		// We only ever send 100 mutations at once. Any more than this and
+		// we risk a timeout from GitHub.
+		for (let i = 0; i < allFilenames.length; i += this.BATCH_SIZE) {
+			const batch = allFilenames.slice(i, i + this.BATCH_SIZE);
+			const mutation = gql`mutation BatchUnmarkFileAsViewedInline {
+				${batch.map((filename, i) =>
+					`alias${i}: ${mutationName}(
+						input: {path: "${filename}", pullRequestId: "${pullRequestId}"}
+					) { clientMutationId }
+					`
+				)}
+			}`;
```

Comment:
Related comment: https://github.com/microsoft/vscode-pull-request-github/pull/4700#discussion_r1165606886

Unfortunately I couldn't find a way around this inlining. It seems when we're creating a query dynamically, this is how it needs to be. Very happy to be shown a different way though; inlining doesn't feel great.